### PR TITLE
Handle full URLs, ip addresses and support Python versions 3.7-3.10

### PR DESCRIPTION
**Pull request by elliotwutingfeng**
# Improvements made

- [x] Support Python versions 3.7-3.10
- [x] Ported over some regex pre-processing code from [tldextract](https://github.com/john-kurkowski/tldextract) to support parsing of full URLs and IP addresses (making this module almost a perfect drop-in replacement for tldextract). Still greatly outperforms **tldextract** and **tld**.
- [x] Renamed `input` variable to `raw_url` because `input` is a Python keyword
- [x] replace redundant `.update()` in `_trie_construct`
- [x] add missing `.decode()` in `format()` to prevent type errors
- [x] rectified bare Exceptions (to prevent linter from complaining)
- [x] fix wrong license in setup.py. Update package description.
- [x] add relevant unit test cases
- [x] ran longer performance benchmarks on faster hardware
- [x] Rephrased wordings in readme, added MIT license badge, and updated performance benchmark values.
- [x] Fix Python 2 compatibility and shell injection prevention (security)

- [x] Trie now supports IDN tlds in unicode form
  ```python
  t.extract("好.地方.新加坡")
  # Before: ('', '', '', '', '', '', '', '')
  # After: ('', '', '好', '地方', '新加坡', '', '', '地方.新加坡')
  ```

- [x] fix `nested_dict` producing erroneous entries.

- [x] add support for extraction of url scheme, userinfo, port, and path
---

Dear jophy,

Many thanks for your innovative use of the trie data structure. I've made some modifications to **fasttld** so that it can be used as a near drop-in replacement for [tldextract](https://github.com/john-kurkowski/tldextract) while preserving existing behaviour for backward-compatibility.

Most notably, the **fasttld** module can now parse full URLs with [schemes, ports, paths, query and fragment](https://en.wikipedia.org/wiki/Uniform_Resource_Identifier#Syntax), thanks to code ported over from [tldextract](https://github.com/john-kurkowski/tldextract)

**fasttld** still greatly outperforms the other modules, despite the additional regex-based pre-processing step, though I have yet to test if my changes have any negative impact on memory usage.

All test cases passed, including the newly added ones.

Let me know what you think!

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,35 +1,42 @@
...
 pip install fasttld
 ```
-or
+
+or build from source
+
 ```python
 git clone https://github.com/jophy/fasttld.git && cd fasttld
-python setup.py
+python setup.py install
```
Comment (elliotwutingfeng) [inline README.md:35]: Previous command doesn't install anything.

File: README.md [hunk 2]
```diff
@@ -1,35 +1,42 @@
-fasttld
-=======
+# fasttld
+
 [![PyPI version](https://badge.fury.io/py/fasttld.svg)](https://badge.fury.io/py/fasttld)
 [![Build Status](https://api.travis-ci.org/jophy/fasttld.svg?branch=master)](https://travis-ci.org/jophy/fasttld)
+[![License](https://img.shields.io/badge/license-MIT-brightgreen)](LICENSE)
```
Comment (elliotwutingfeng) [inline README.md:5]: License badge

File: README.md [hunk 3]
```diff
@@ -1,35 +1,42 @@
...
 
-`Fasttld` is a high performance TLD extract module based on a compressed [trie](https://en.wikipedia.org/wiki/Trie) 
-with builtin python dict().
+**fasttld** is a high performance [top level domains (TLD)](https://en.wikipedia.org/wiki/Top-level_domain) extraction module based on the compressed [trie](https://en.wikipedia.org/wiki/Trie) data structure
```
Comment (elliotwutingfeng) [inline README.md:7]: rephrasing

File: README.md [hunk 4]
```diff
@@ -1,35 +1,42 @@
...
 
 ![Trie](https://upload.wikimedia.org/wikipedia/commons/b/be/Trie_example.svg)
 
-# Background
-The plan of fasttld module is to extract top level domains from millions lines of domains (from DNS data) in one time. 
-In the other words, we extract "www.google.com" into "google.com". 
- 
-Most programmers think it is easy, just run `".".join(domain.split('.')[1::])`. But it is wrong ! Think about how to 
-process "www.baidu.com.cn". So we must know what suffixes are.
+## Background
```
Comment (elliotwutingfeng) [inline README.md:12]: rephrasing

File: README.md [hunk 5]
```diff
@@ -1,35 +1,42 @@
...
+
+The goal of **fasttld** is to extract [top level domains (TLDs)](https://en.wikipedia.org/wiki/Top-level_domain) from [URLs](https://en.wikipedia.org/wiki/URL) efficiently. In the other words, we extract `com` from URLs like `www.google.com` or `https://maps.google.com:8080/a/long/path/?query=42`.
```
Comment (elliotwutingfeng) [inline README.md:14]: rephrasing

File: README.md [hunk 6]
```diff
@@ -46,71 +53,96 @@ python setup.py
...
 'google.com'
 ```
-extract() returns a tupple (subdomain, domain, suffix, domain_name) .
 
-# Update Public Suffix List
-`fasttld` will update Public Suffix List every 3 days automatically when it is called.
-You can also do update manually. Try the following commands.
+extract() returns a tuple `(subdomain, domain, suffix, domain_name)` .
+
+## Update the Mozilla Public Suffix List local copy
```
Comment (elliotwutingfeng) [inline README.md:51]: rephrasing

File: README.md [hunk 7]
```diff
@@ -46,71 +53,96 @@ python setup.py
...
 >>> from fasttld import FastTLDExtract
->>> FastTLDExtract().extract('domain', subdomain=False)
+>>> FastTLDExtract().extract('domain', subdomain=False) # set subdomain=False
 ```
-# Exclude private domain
-Due to security issues, public suffix list contains some private domains. Eg, blogspot.co.uk, sinaapp.com.
 
-`fasttld` can exclude private domain at the beginning of constructing the suffix trie. Samples see bellow.
+## Optional: Exclude private domains
+
```
Comment (elliotwutingfeng) [inline README.md:87]: Added detailed explanation on why private domains exist in the PSL.

File: README.md [hunk 8]
```diff
@@ -46,71 +53,96 @@ python setup.py
...
+Python 3.9.12, AMD Ryzen 7 5800X 3.8 GHz 8 cores 16 threads, 48GB RAM
 
-Initialize the class just one time, calling extract function one million times. Results see below.
+### Test results
 
-Test environment: Macbook Pro 13', Intel Core i5 2.7 GHz, 8GB RAM.
+| **module\case** | `jophy.com` | `www.baidu.com.cn` | `jo.noexist` | `https://maps.google.com.ua/a/long/path?query=42` | `1.1.1.1` | `https://192.168.55.1` |
+|-----------------|---------------|----------------------|----------------|-----------------------------------------------------|-------------|--------------------------|
+| fasttld         | 7.60s         | 9.90s                | 5.28s          | 5.67s                                               | 5.06s       | 5.30s                    |
```
Comment (elliotwutingfeng) [inline README.md:124]: Added longer tests (10 million entries) on my faster hardware.

File: README.md [hunk 9]
```diff
@@ -46,71 +53,96 @@ python setup.py
...
 
-Disable subdomain output:
+| **module\case** | `jophy.com` | `www.baidu.com.cn` | `jo.noexist` | `https://maps.google.com.ua/a/long/path?query=42` | `1.1.1.1` | `https://192.168.55.1`
+|-----------------|---------------|----------------------|----------------|-----------------------------------------------------|-------------|--------------------------|
+| fasttld         | 7.55s         | 8.98s                | 5.20s          | 5.52s                                               | 5.13s       | 5.25s                    |
 
- module\case | jophy.com | www.baidu.com.cn|jo.noexist
--------------|-----------|-----------------|----------
-fasttld      |    2.73   |       3.08      |  1.90
+On average, **fasttld** is **4 to 5** times faster than the other modules. It retains its performance advantage even when parsing long URLs like `https://maps.google.com.ua/a/long/path?query=42`
```
Comment (elliotwutingfeng) [inline README.md:136]: Performance advantage is still about 5 times! :smile:

File: fasttld/FastTLDExtract.py [hunk 10]
```diff
@@ -2,17 +2,42 @@
...
 """
-from fasttld.psl import getPublicSuffixList
-from fasttld.psl import update
+import re
+import socket
+from operator import itemgetter
 
 import idna
 
+from fasttld.psl import getPublicSuffixList, update
```
Comment (elliotwutingfeng) [inline fasttld/FastTLDExtract.py:16]: fasttld can now check for ip addresses

File: fasttld/FastTLDExtract.py [hunk 11]
```diff
@@ -24,21 +49,35 @@ def update(self, *args, **kwargs):
     def nested_dict(self, dic, keys):
         """
         The idea of this function is based on https://stackoverflow.com/questions/13687924
-        :param dic: 
-        :param keys: 
-        :return: 
+
```
Comment (elliotwutingfeng) [inline fasttld/FastTLDExtract.py]: Explains how the nested_dict() function works

File: fasttld/FastTLDExtract.py [hunk 12]
```diff
@@ -24,21 +49,35 @@ def update(self, *args, **kwargs):
...
+        :param dic:
+        :param keys:
+        :return:
         """
-        for key in keys[:-1]:
+        keys_except_last_key, second_last_key, last_key = itemgetter(slice(0, -1), -2, -1)(keys)
+        for key in keys_except_last_key:
             dic_bk = dic
-            dic = dic.setdefault(key, {})
```
Comment (elliotwutingfeng) [inline fasttld/FastTLDExtract.py:33]: replaced setdefault and update for a slight performance boost.

File: fasttld/FastTLDExtract.py [hunk 13]
```diff
@@ -49,84 +88,83 @@ def _trie_construct(self, exclude_private_suffix, file_path=''):
...
-        else:
-            SuffixList = AllSuffixList
+        SuffixList = PublicSuffixList if exclude_private_suffix else AllSuffixList
         for suffix in SuffixList:
             if '.' in suffix:
                 sp = suffix.split('.')
                 sp.reverse()
                 self.nested_dict(tld_trie, sp)
             else:
-                tld_trie.update({
```
Comment (elliotwutingfeng) [inline fasttld/FastTLDExtract.py:62]: replaced update() with equivalent faster code

File: fasttld/FastTLDExtract.py [hunk 14]
```diff
@@ -49,84 +88,83 @@ def _trie_construct(self, exclude_private_suffix, file_path=''):
...
-                    suffix: {
-                        '_END': True
-                    }
-                })
+                tld_trie[suffix] = {'_END': True}
         for key, val in tld_trie.items():
-            if len(tld_trie[key]) == 1:
```
Comment (elliotwutingfeng) [inline fasttld/FastTLDExtract.py:68]: replaced update() with equivalent faster code

File: fasttld/FastTLDExtract.py [hunk 15]
```diff
@@ -49,84 +88,83 @@ def _trie_construct(self, exclude_private_suffix, file_path=''):
...
-                    })
+            if len(val) == 1 and '_END' in val:
+                tld_trie[key] = True
         return tld_trie
-    
+
     def __call__(self, *args, **kwargs):
         return self.extract(*args, **kwargs)
 
-    def extract(self, input, subdomain=True, format=False):
```
Comment (elliotwutingfeng) [inline fasttld/FastTLDExtract.py:78]: replaced use of input name as a variable, as input() is a reserved python keyword.

File: fasttld/FastTLDExtract.py [hunk 16]
```diff
@@ -49,84 +88,83 @@ def _trie_construct(self, exclude_private_suffix, file_path=''):
...
         """
-        ret_subdomain = ''
-        ret_domain = ''
-        ret_suffix = ''
-        ret_domain_name = ''
+        ret_subdomain = ret_domain = ret_suffix = ret_domain_name = ''
         if format:
-            input = self.format(input)
-        labels = input.split('.')
+            raw_url = self.format(raw_url)
```
Comment (elliotwutingfeng) [inline fasttld/FastTLDExtract.py:117]: fasttld can now parse full urls that include the scheme (e.g. https:// ftp:// etc.) and anything after the host subcomponent (e.g. :5000/index.html?id=42 etc.)

File: fasttld/FastTLDExtract.py [hunk 17]
```diff
@@ -49,84 +88,83 @@ def _trie_construct(self, exclude_private_suffix, file_path=''):
...
+        )
+        # Determine if raw_url is an IP address
+        if len(netloc) != 0 and looks_like_ip(netloc):
+            return ("", netloc, "", netloc)
+        labels = netloc.split(".")
         labels.reverse()
         node = self.trie  # define the root node
         suffix = []
-
-        for index, label in enumerate(labels):
```
Comment (elliotwutingfeng) [inline fasttld/FastTLDExtract.py:99]: We don't actually need the index, therefore enumerate was removed.

File: fasttld/FastTLDExtract.py [hunk 18]
```diff
@@ -49,84 +88,83 @@ def _trie_construct(self, exclude_private_suffix, file_path=''):
...
             if node is True:
                 # This node is an end node.
                 ret_domain = label
                 break
 
             # This node has sub-nodes and maybe an end-node.
-            # eg. us -> (us, gov.us)
+            # eg. cn -> (cn, gov.cn)
             if '_END' in node:
-                if index < len(labels):
```
Comment (elliotwutingfeng) [inline fasttld/FastTLDExtract.py:108]: `if index < len(labels)` is redundant here as it is already guaranteed to be `True`.

File: fasttld/FastTLDExtract.py [hunk 19]
```diff
@@ -49,84 +88,83 @@ def _trie_construct(self, exclude_private_suffix, file_path=''):
...
-                    # check if there is a sub node
-                    # eg. www.ck
-                    if '!'+label in node:
-                        ret_domain = label
-                    else:
-                        suffix.append(label)
-                    break
+                # check if there is a sub node
+                # eg. www.ck
+                if ("!%s" % label) in node:
```
Comment (elliotwutingfeng) [inline fasttld/FastTLDExtract.py:195]: slightly faster than string concatenation

File: fasttld/FastTLDExtract.py [hunk 20]
```diff
@@ -140,11 +178,11 @@ def extract(self, input, subdomain=True, format=False):
         len_labels = len(labels)
         ret_suffix = ".".join(suffix)
 
-        if len_suffix < len_labels and len_suffix != 0:
+        if 0 < len_suffix < len_labels:
             ret_domain = labels[len_suffix]
             if subdomain:
                 if len_suffix + 1 < len_labels:
-                    ret_subdomain = input[:-(len(ret_domain + ret_suffix) + 2)]
+                    ret_subdomain = netloc[:-(len(ret_domain) + len(ret_suffix) + 2)]
```
Comment (elliotwutingfeng) [inline fasttld/FastTLDExtract.py]: removed unnecessary string concatenation

File: fasttld/FastTLDExtract.py [hunk 21]
```diff
@@ -154,23 +192,23 @@ def extract(self, input, subdomain=True, format=False):
...
         :return: input
         """
-        input = input.strip().lower()
-        input = idna.encode(input)
-        # input = urlparse.urlparse(input).netloc
-        # if '//' in input:
-        #     _, _, input = input.rpartition('//')
-        # if '/' in input:
-        #     input, _, _ = input.lpartition('//')
+        # idna_url = idna.encode(raw_url.strip().lower()).decode()
```
Comment (elliotwutingfeng) [inline fasttld/FastTLDExtract.py:234]: Added missing `decode()`

File: fasttld/psl.py [hunk 22]
```diff
@@ -32,7 +32,7 @@ def getPublicSuffixList(file_path):
 
     try:
         fd = open(file_path, 'r', encoding='utf-8')
-    except:
+    except Exception:
```
Comment (elliotwutingfeng) [inline fasttld/psl.py:35]: Prevent linter from complaining about bare exceptions

File: fasttld/psl.py [hunk 23]
```diff
@@ -45,7 +45,7 @@ def getPublicSuffixList(file_path):
                 continue
             try:
                 suffix = line.decode('utf-8').encode('idna')
-            except:
+            except Exception:
```
Comment (elliotwutingfeng) [inline fasttld/psl.py:50]: Prevent linter from complaining about bare exceptions

File: fasttld/psl.py [hunk 24]
```diff
@@ -68,11 +68,11 @@ def update(show_output=True):
             import urllib
             downfile = urllib.URLopener()
             downfile.retrieve(base_url, file_path)
-        except:
+        except Exception:
```
Comment (elliotwutingfeng) [inline fasttld/psl.py:78]: Prevent linter from complaining about bare exceptions

File: setup.py [hunk 25]
```diff
@@ -1,39 +1,41 @@
...
+
+version = 'v0.2.2'
 
 
 def test_suite():
     try:
         import unittest2
         unittest = unittest2
-    except:
+    except Exception:
```
Comment (elliotwutingfeng) [inline setup.py:13]: Prevent linter from complaining about bare exceptions

File: setup.py [hunk 26]
```diff
@@ -1,39 +1,41 @@
...
     test_loader = unittest.TestLoader()
     test_suite = test_loader.discover('tests', pattern='maintest.py')
     return test_suite
 
+
 if sys.argv[-1] == 'publish':
-    os.system("python setup.py sdist upload")
-    os.system("git tag v%s" % (version))
-    os.system("git push --tags")
+    subprocess.call(["python", "setup.py", "sdist", "upload"])
```
Comment (elliotwutingfeng) [inline setup.py:22]: Prevent shell injection.

File: setup.py [hunk 27]
```diff
@@ -1,39 +1,41 @@
...
     sys.exit()
 
 
 setup(
     name='fasttld',
-    version=version,
+    version=version[1:],
     packages=['fasttld'],
     url='https://github.com/jophy/fasttld',
-    license='GPL',
```
Comment (elliotwutingfeng) [inline setup.py:33]: Updated license to MIT

File: tests/maintest.py [hunk 28]
```diff
@@ -1,6 +1,7 @@
-from fasttld import FastTLDExtract
```
Comment (elliotwutingfeng) [inline tests/maintest.py:1]: Added more test cases to cover new features as well as existing features.

File: tests/performance.py [hunk 29]
```diff
@@ -2,96 +2,55 @@
...
-for i in range(1, 1000000):
-    get_tld(case1, fix_protocol=True)
-print("tld in case2: %ss" % (time.time() - t1))
 
-t1 = time.time()
-for i in range(1, 1000000):
-    get_tld(case1, fix_protocol=True)
-print("tld in case3: %ss\n" % (time.time() - t1))
\ No newline at end of file
+cases = [
```
Comment (elliotwutingfeng) [inline tests/performance.py:20]: Added more test cases for performance benchmarks.

File: tests/performance.py [hunk 30]
```diff
@@ -2,96 +2,55 @@
...
+         'jophy.com',
+         'www.baidu.com.cn',
+         'jo.noexist',
+         'https://maps.google.com.ua/a/long/path?query=42',
+         '1.1.1.1', 'https://192.168.1.1'
+        ]
+
+if sys.version_info.major == 2:
```
Comment (elliotwutingfeng) [inline tests/performance.py:28]: Previous python 2 compatibility check wasn't working in Python 2.

File: tests/performance.py [hunk 31]
```diff
@@ -2,96 +2,55 @@
...
+              {}),
+              ('tld',
+              get_tld,
+              {'fix_protocol': True, 'fail_silently': True}),
+              ('fasttld_without_subdomains',
+               t.extract,
+              {'subdomain': False})
+              ]
+
+for test_case in test_cases:
```
Comment (elliotwutingfeng) [inline tests/performance.py:49]: Put everything into a loop to reduce code duplication

File: tests/performance.py [hunk 32]
```diff
@@ -2,96 +2,55 @@
...
+    module, extractor, kwargs = test_case
+    for url in cases:
+        t1 = time.perf_counter()
```
Comment (elliotwutingfeng) [inline tests/performance.py:52]: replaced time.time() with time.perf_counter()

**Comment by jophy** [comment]
Thanks a lot.

I will review all ASAP.

File: fasttld/FastTLDExtract.py (hunk 11 above)
Comment (elliotwutingfeng) [reply fasttld/FastTLDExtract.py]: wrong explanation, I misunderstood how it worked!

File: fasttld/FastTLDExtract.py [hunk 33]
```diff
@@ -24,23 +48,24 @@ def update(self, *args, **kwargs):
...
         """
+        end = False
         for key in keys[:-1]:
             dic_bk = dic
-            dic = dic.setdefault(key, {})
+            if key not in dic:
+                dic[key] = {}
+            dic = dic[key]
             if isinstance(dic, bool):
+                end = True
```
Comment (elliotwutingfeng) [inline fasttld/FastTLDExtract.py:70]: Without this `end=True` flag, the test case in `test_nested_dict()` will have the dictionary include an erroneous extra `com: True`

File: fasttld/FastTLDExtract.py (hunk 16 above)
Comment (jophy) [reply fasttld/FastTLDExtract.py:117]: When I first built the fasttld python module, it was used for processing large dns datasets.

URL should be parsed before call `extract`.

I will merge your great work and make some improvements.

**Review by jophy — APPROVED** [review]

**Pull request merged**